File: Common/Network/Message.cs
using System.Text;

namespace CommonData.Network;

public abstract class Message
{
    public abstract MessageHeader GetHeader();

    public Task WriteDataTo(Stream targetStream)
    {
        using Stream source = GetData();
        source.Flush();
        source.Position = 0;
        byte[] buffer = new byte[source.Length];
        source.Read(buffer, 0, buffer.Length);
        return targetStream.WriteAsync(buffer, 0, buffer.Length);
    }

    public abstract void ApplyData(Stream stream);

    protected abstract Stream GetData();
    
    protected byte[] ByteArrayReadStream(Stream stream, int arrayLength)
    {
        byte[] bytes = new byte[arrayLength];
        stream.Read(bytes, 0, arrayLength);
        return bytes;
    }

    protected String StringReadStream(Stream stream)
    {
        int stringLength = IntReadStream(stream);
        byte[] stringBuffer = new byte[stringLength];
        stream.Read(stringBuffer, 0, stringLength);
        return Encoding.UTF8.GetString(stringBuffer);
    }

    protected int IntReadStream(Stream stream)
    {
        byte[] intBuffer = new byte[4];
        stream.Read(intBuffer, 0, 4);
        return BitConverter.ToInt32(intBuffer, 0);
    }

    protected bool BoolReadStream(Stream stream)
    {
        byte[] byteBuffer = new byte[1];
        stream.Read(byteBuffer, 0, 1);
        return BitConverter.ToBoolean(byteBuffer);
    }

    protected void WriteToStream(Stream stream, string value)
    {
        byte[] stringBytes = Encoding.UTF8.GetBytes(value);
        stream.Write(BitConverter.GetBytes(stringBytes.Length));
        stream.Write(stringBytes);
    }

    protected void WriteToStream(Stream stream, int value)
    {
        stream.Write(BitConverter.GetBytes(value));
    }

    protected void WriteToStream(Stream stream, bool value)
    {
        stream.Write(BitConverter.GetBytes(value));
    }
}

File: IntegrationMod/src/main/java/com/example/examplemod/bfml/Message.java
package com.example.examplemod.bfml;

import org.apache.http.MethodNotSupportedException;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;

public abstract class Message {

    public Message() {

    }

    public abstract MessageHeader getMessageHeader() throws Exception;

    public abstract void applyData(InputStream stream) throws IOException, MethodNotSupportedException;

    public abstract byte[] getData() throws Exception;

    public void writeDataTo(OutputStream targetStream) throws Exception {
        targetStream.write(getData());
    }

    protected byte[] byteArrayReadStream(InputStream stream, int arrayLength) throws IOException {
        byte[] bytes = new byte[arrayLength];
        stream.read(bytes, 0, arrayLength);
        return bytes;
    }

    protected String stringReadStream(InputStream stream) throws IOException {
        int stringLength = intReadStream(stream);
        byte[] stringBuffer = new byte[stringLength];
        stream.read(stringBuffer, 0, stringLength);
        ByteBuffer buffer = ByteBuffer.wrap(stringBuffer);
        buffer.order(ByteOrder.LITTLE_ENDIAN);
        byte[] bufferBytes = buffer.array();
        String bufferString = new String(bufferBytes, StandardCharsets.UTF_8);
        return bufferString;
    }

    protected int intReadStream(InputStream stream) throws IOException {
        byte[] intBuffer = new byte[4];
        stream.read(intBuffer, 0, 4);
        ByteBuffer buffer = ByteBuffer.wrap(intBuffer);
        buffer.order(ByteOrder.LITTLE_ENDIAN);
        return buffer.getInt();
    }

    protected boolean boolReadStream(InputStream stream) throws IOException {
        byte[] byteBuffer = new byte[1];
        stream.read(byteBuffer, 0, 1);
        return (int) byteBuffer[0] != 0;
    }
}
